packages/worker/src/worker-tuner.ts

```diff
@@ -73,15 +87,127 @@ export interface ResourceBasedTuner {
+
+export interface WorkflowSlotInfo {
+  type: 'workflow';
+  workflowId: string;
+  runId: string;
+}
+
+export interface ActivitySlotInfo {
+  type: 'activity';
+  activityId: string;
+}
+
+export interface LocalActivitySlotInfo {
+  type: 'local-activity';
+  activityId: string;
+}
+
+export interface UserSlotSupplier<SI extends SlotInfo> {
+  type: 'user';
+
+  /**
+   * This function is called before polling for new tasks. Your implementation should block until a
+   * slot is available then return a permit to use that slot.
+   *
+   * // TODO: How to handle cancellation? AbortController? CancelToken?
+   *
+   * @param ctx The context for slot reservation.
+   * @returns A permit to use the slot which may be populated with your own data.
+   */
+  reserveSlot(ctx: SlotReserveContext): Promise<SlotPermit>;
+
+  /**
+   * This function is called when trying to reserve slots for "eager" workflow and activity tasks.
+   * Eager tasks are those which are returned as a result of completing a workflow task, rather than
+   * from polling. Your implementation must not block, and If a slot is available, return a permit
+   * to use that slot.
+   *
+   * @param ctx The context for slot reservation.
+   * @returns Maybe a permit to use the slot which may be populated with your own data.
+   */
+  tryReserveSlot(ctx: SlotReserveContext): SlotPermit | undefined;
+
+  /**
+   * This function is called once a slot is actually being used to process some task, which may be
+   * some time after the slot was reserved originally. For example, if there is no work for a
+   * worker, a number of slots equal to the number of active pollers may already be reserved, but
+   * none of them are being used yet. This call should be non-blocking.
+   *
+   * @param ctx The context for marking a slot as used.
+   */
+  markSlotUsed(slot: SlotMarkUsedContext<SI>): void;
+
+  /**
+   * This function is called once a permit is no longer needed. This could be because the task has
+   * finished, whether successfully or not, or because the slot was no longer needed (ex: the number
+   * of active pollers decreased). This call should be non-blocking.
+   *
+   * @param ctx The context for releasing a slot.
+   */
+  releaseSlot(slot: SlotReleaseContext<SI>): void;
+}
+
+export interface SlotPermit {
+  // Use this field to associate any data you like with a particular slot
+  data?: any;
+}
+
+export interface SlotReserveContext {
+  // The name of the task queue for which this reservation request is associated
+  taskQueue: string;
+  // The build id of the worker that is requesting the reservation
+  workerIdentity: string;
+  // The identity of the worker that is requesting the reservation
+  workerBuildId: string;
+  // True iff this is a reservation for a sticky poll for a workflow task
+  isSticky: boolean;
+
+  // Returns the number of currently outstanding slot permits, whether used or un-used.
+  numIssuedSlots(): number;
```

Comment:
It's because this value could change over the course of a call to reserve a slot if that call is awaiting on other stuff.

Reply:
But doesn't the provider itself already know about the current number of slots it allowed? Why do we even need Core to provide that information?

Reply:
Oh - yes. It could technically know that. Just in practice this is probably something you pretty commonly want and it's nice to have handled for you. That said I'm fine w/ leaving it off initially and adding it in if requested.

Reply:
I'd rather leave it out for now, and see while progressing either that's useful.

I foresee some potential synchronization issues that users may struggle with; I can think this more in depth if needed, but my intuition is that anyway, this function would rarely be pertinent in anything more than a trivial sample `SlotSupplier`.